### PR TITLE
APIS-5615 - Decline journey and submission shape change

### PR DESCRIPTION
**Pull request by AndySpaven**

### CODE REVIEW

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/ConfirmYourDecisionController.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  import ConfirmYourDecisionController._
+
+  def page(applicationId: ApplicationId): Action[AnyContent] = loggedInWithApplicationAndSubmission(applicationId) { implicit request =>
+    successful(Ok(confirmYourDecisionPage(ViewModel(applicationId, request.application.name))))
+  }
+
+  def action(applicationId: ApplicationId): Action[AnyContent] = loggedInWithApplicationAndSubmission(applicationId) { implicit request => 
+    request.body.asFormUrlEncoded.getOrElse(Map.empty).get("grant-decision").flatMap(_.headOption) match {
+      case Some("decline")              => successful(Redirect(uk.gov.hmrc.apigatekeeperapprovalsfrontend.controllers.routes.DeclinedJourneyController.provideReasonsPage(applicationId)))
+      // TODO
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/ConfirmYourDecisionController.scala:76]: is this needed?

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/actions/ApplicationActionBuilders.scala [hunk 2]
```diff
@@ -51,7 +51,10 @@ trait ApplicationActionBuilders {
         import cats.implicits._
 
         applicationActionService.process(applicationId, request)
-        .toRight(NotFound(errorHandler.notFoundTemplate(Request(request, request.messagesApi)))).value
+        .toRight( {
+          println("Oh dear")
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/actions/ApplicationActionBuilders.scala:55]: let's remove this

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ApplicationApprovedPage.scala.html [hunk 3]
```diff
@@ -14,22 +14,33 @@
...
+        @messages("applicationapproved.productionaccessgranted")
+      </h1>
+      <div class="govuk-panel__body">
+        @model.appName
+      </div>
+    </div>
+    <br>
+
+    <p class="govuk-body">
+      <a class="govuk-link govuk-link--no-visited-state" href="@{s"${config.applicationsPageUri}"}">Check another production credentials request.</a>
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ApplicationApprovedPage.scala.html:42]: this text should be in the messages file

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ApplicationDeclinedPage.scala.html [hunk 4]
```diff
@@ -24,11 +24,24 @@
...
-  <div class="govuk-box-highlight">
-    <h1 class="govuk-heading-l govuk-highlight-text">@{messages("applicationdeclined.title")}</h1>
-    <p class="govuk-body govuk-highlight-text">@{messages("applicationdeclined.text")}</p>
-  </div>
+  <div class="govuk-width-container">
+    <div class="govuk-grid-row">
+      <div class="govuk-grid-column-three-quarters">
+        <div class="govuk-panel govuk-panel--confirmation">
+          <h1 class="govuk-panel__title">
+            You have declined this request
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ApplicationDeclinedPage.scala.html:33]: messages file

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ApplicationDeclinedPage.scala.html [hunk 5]
```diff
@@ -24,11 +24,24 @@
...
+          </h1>
+          <div class="govuk-panel__body">
+            @model.appName
+          </div>
+        </div>
+        <br>
 
+        <p class="govuk-body">
+          <a class="govuk-link govuk-link--no-visited-state" href="../../index-results">Check another production credentials request.</a>
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ApplicationDeclinedPage.scala.html:42]: messages file

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ConfirmYourDecisionPage.scala.html [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    
+      </fieldset>
+    </div>
+
+    <div class="govuk-button-group">
+      <button id="checked" class="govuk-button govuk-button" type="submit" name="submit-action">@{messages("confirmyourdecision.continue")}</button>
+    </div>
+
+    <p class="govuk-body">
+      You can escalate the request to a senior colleague if you need to. <a class="govuk-link govuk-link--no-visited-state" href="#">Check the guidance on Confluence</a>.
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ConfirmYourDecisionPage.scala.html:67]: broken confluence link

Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ConfirmYourDecisionPage.scala.html:67]: should be in messages file

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ProvideReasonsForDecliningPage.scala.html [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+@layout(pageTitle = Some(messages("applicationdeclined.pagetitle")), beforeContent = Some(breadcrumbs(None))) {
+  <link rel="stylesheet" href='@routes.Assets.versioned("stylesheets/application.css")'>
+  
+  @helper.form(action = uk.gov.hmrc.apigatekeeperapprovalsfrontend.controllers.routes.DeclinedJourneyController.provideReasonsAction(model.applicationId), 'class -> "form") {
+    @helper.CSRF.formField
+
+    <span class="govuk-caption-l">@model.appName</span>
+    <h1 class="govuk-label-wrapper"><label class="govuk-label govuk-label--l" for="reasons">
+        Note all the reasons the request is declined
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/views/ProvideReasonsForDecliningPage.scala.html:33]: a number of lines of hard-coded text

File: app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala [hunk 8]
```diff
@@ -67,15 +66,81 @@ object Submission {
...
+    s.copy(instances = NonEmptyList(delta(s.instances.head), s.instances.tail))
+  }
+
+  val addStatusHistory: (Submission.Status) => Submission => Submission = newStatus => s => {
+    require(Submission.Status.isLegalTransition(s.status, newStatus))
+    changeLatestInstance(_.copy(statusHistory = newStatus :: s.latestInstance.statusHistory))(s)
+  }
+
+  val changeStatusHistory: (Submission.Status => Submission.Status) => Submission => Submission = delta => s => {
+    val inStatus = s.latestInstance.statusHistory.head
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala:97]: ditto other reviews - should we check this makes for a legal transition?

File: test/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/AbstractControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    val strideAuthConfig = app.injector.instanceOf[StrideAuthConfig]
+    val forbiddenHandler = app.injector.instanceOf[HandleForbiddenWithView]
+    val mcc = app.injector.instanceOf[MessagesControllerComponents]
+    val errorHandler = app.injector.instanceOf[ErrorHandler]
+
+    val application = anApplication(applicationId)
+    
+    val fakeRequest = FakeRequest().withCSRFToken
+
+    val fakeSubmitCheckedRequest = FakeRequest().withCSRFToken.withFormUrlEncodedBody("submit-action" -> "checked")
```
Comment (robdawson-hmrc) [inline test/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/AbstractControllerSpec.scala:59]: could re-use `fakeRequest` on these lines

File: app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala (hunk 8 above)
Comment (AndySpaven) [reply app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala:97]: Fix from TPA

File: test/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/AbstractControllerSpec.scala (hunk 9 above)
Comment (AndySpaven) [reply test/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/AbstractControllerSpec.scala:59]: Done

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/actions/ApplicationActionBuilders.scala (hunk 2 above)
Comment (AndySpaven) [reply app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/actions/ApplicationActionBuilders.scala:55]: Done

File: app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/ConfirmYourDecisionController.scala (hunk 1 above)
Comment (AndySpaven) [reply app/uk/gov/hmrc/apigatekeeperapprovalsfrontend/controllers/ConfirmYourDecisionController.scala:76]: Yes, in so much as we need to play the next ticket for this.

**Review by robdawson-hmrc — APPROVED** [review]

**Pull request merged**